termqt/terminal_io_windows.py

```diff
@@ -0,0 +1,107 @@
+import os
+import logging
+import threading
+import winpty
+
+
+class TerminalIO:
+    # This class provides io functions that communciate with the Terminal
+    # and the pty (pseudo-tty) of a program.
+
+    def __init__(self, cmd: str, cols: int, rows: int, logger=None):
+        # Initilize.
+        #
+        # args: cols: columns
+        #       rows: rows
+        #       cmd: the command to execute.
+        #       env: environment variables, if None, set it to os.environ
+        self.cmd = cmd
+        self.logger = logger if logger else logging.getLogger()
+        self.cols = cols
+        self.rows = rows
+        self.running = False
+
+        self._read_buf = b""
+
+        self.terminated_callback = lambda: None
+        self.stdout_callback = lambda bs: None
+
+    def spawn(self):
+        # Spawn the sub-process in pty.
+        try:
+            self.running = True
+            self.pty_process = winpty.PtyProcess.spawn(
+                self.cmd,
+                dimensions=(self.rows, self.cols),
+                backend=1,
+            )
+            self.pty_process.setwinsize(self.cols, self.rows)
+            
+            self._read_thread = threading.Thread(
+                name="TerminalIO Read Loop",
+                target=self._read_loop,
+                daemon=True
+            )
+            self._read_thread.start()
+        except:
+            self.running = False
+            self.terminated_callback()
+
+    def resize(self, rows, cols):
+        try:
+            self.cols = cols
+            self.rows = rows
+            self.pty_process.setwinsize(self.rows, self.cols)
+        except (OSError, AssertionError):
+            self.running = False
+            self.terminated_callback()
+
+    def write(self, buffer: bytes):
+        self.logger.debug("stdin: " + str(buffer))
+        if not self.running:
+            return
+        try:
+            self.pty_process.write(buffer.decode("utf-8"))
+        except (OSError, AssertionError):
+            self.running = False
+            self.terminated_callback()
+
+    def _read_loop(self):
+        # read loop to be run in a separated thread
+        try:
+            while self.running:
+                buf = self.pty_process.read()
```

Comment:
For the issue mentioned in #5, could you check how many characters you get every time by this `read` function?

Reply:
As I mentioned in the #5 , this slow rendering is present even before this pull request, on the `master` branch.